Infrastructure docs (#108)
populationgenomics/team-docs: infrastructure.md

# CPG Infrastructure

## Dataset

> A dataset is one defined collection of resources. It has:

Reviewer: ```suggestion
A *dataset* is a defined collection of resources, with an associated permissions group for controlling access. It has:
```

- A GCP project

> - A set of buckets

Reviewer: ```suggestion
- A set of GCS buckets
```

- A set of secrets to manage access (access_group_cache, cromwell access keys)

> - A set of allowed_repositories

Reviewer: ```suggestion
- A set of GitHub repositories, whose code is allowed to access this data
```

> ## Components

Reviewer: Cromwell might be worth mentioning here, too, particularly since long term we'd like to move to Terra for that functionality.

> - Hail query / batch

Reviewer: ```suggestion
- [Hail Batch](https://github.com/populationgenomics/hail)
- Hail Query (currently through Dataproc)
```

- Analysis-runner

> - Sample-metadata server

Reviewer: Could you add the repo links here and add a short summary of what these are used for?

## Permissions

The [storage policies](storage_policies) document has more information about the different ways we store data in the CPG, but broadly, each dataset has three levels:

- `test`: can run any commit on the allowed repositories, on the `dataset-test*` buckets
- `standard`: commit MUST be on the main branch, allowed READ / WRITE access to `dataset-main*` buckets
- `full`: commit MUST be on the branch branch, allowed all operations on `dataset-main*` and `dataset-test` buckets

## Example: seqr loading pipeline

The seqr loading pipeline is good example, as it uses almost every single piece of architecture that the CPG offers.

This relies on:

- samples being loaded into a `dataset-main*` bucket
- sample metadata being loaded into the sample-metadata system

>     - Including pedigree data, individual level metadata (eg: HPO terms)

Reviewer: This is kind of vague -- can you outline here what the soonish picture of seqr wizard + sample metadata API interactions look like? For someone not familiar with our process, it's unclear what "loaded" means in this context.

Entrypoint: [GH: hail-elasticsearch-pipelines::batch_seqr_loader/batch_workflow.py](https://github.com/populationgenomics/hail-elasticsearch-pipelines/blob/main/batch_seqr_loader/batch_workflow.py)

1. The batch_workflow script is submitted to the analysis-runner, to use `main` data
1. This script first contacts the sample-metadata system to get the active samples for the current project.

> 1. It uses state in the sample-metadata to work out which samples still need to realigned, and variant called

Reviewer: Worth mentioning here that this happens all within Hail Batch, through an adapted translation of the WARP pipeline.

I'd also mention GATK-SV here, as it's adds substantial complexity (calling into Cromwell from Hail Batch).

> 1. After potential alignment and variant calling, it determines the samples to add to the joint-call

Reviewer: "determines" is a bit vague here. This is based on what's stored in the sample metadata, right?

> 1. Spin up a dataproc cluster with vep, and submit `batch_seqr_loader/scripts/load_project_to_es.py` to the dataproc cluster.

Reviewer: Is it worth splitting these points into stages and explaining what the system interactions are in each stage?
- Validation (e.g. relatedness)
- Alignment
- Variant calling
- SV calling
- Joint calling / filtering
- Annotation
- Export to Elasticsearch

Does this documentation already exist as part of the loading pipeline? If not, maybe we should just add this documentation there and link to it here?

> 1. The `load_project_to_es` workflow uses hail to annotate the joint-call set with VEP, and a number of other clinical databases, before loading into elasticsearch.

Reviewer: Might be worth mentioning that we run Managed Elasticsearch (through the GCP Marketplace).

Now that the elastic search index has been created:

- The seqr project can be created (if not already),
- Export the following from the sample-metadata server and load into seqr:
    - Pedigree
    - (_in-progress_) family phenotypes
    - Individual ID,Cram path,Internal SampleID map
    - Participant-id to internal sample id map, import with the ES index name

>     - Individual level metadata

Reviewer: That's probably too detailed. I wouldn't worry so much about the individual steps to set up a project here, than describing how different parts of the system architecture interact.

I.e. what's interesting here is that there's a publicly accessible web frontend (behind IAP) that uses Google OAuth2 for authentication. It queries the previously created Elasticsearch index and raw reads from the GCP buckets.
